### PR TITLE
Add AutoAugment and ImageNet policy

### PR DESCRIPTION
**Pull request by stiepan**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Adds AutoAugment: augmentation scheme that applies augmentations randomly per sample based on a policy. The policy consits of sub-polcies, each sub-policy is a sequence of augmentations to apply. Sub-policies are chosen uniformy at random for each sample. Then, the augmentations from the sequence are applied one-by-one. Each augmentation is described with the probability of it being applied or skipped and a magnitude - inuitevely a strenght of the operation.

The ImageNet policy is a policy found by the authors of the paper that introduces AA shown to improve the accuracy (https://arxiv.org/abs/1805.09501).

## Additional information:
This is a part of Automatic Augmentation module https://github.com/NVIDIA/DALI/pull/4648

The implementation groups the same augmentations from different sub-policies in hope for bigger partial-batches. For example, the ImageNet policy has 25 sub-policies, but in fact only 8 augmentations at each stage of the sub-policy.

Take the policy:
```
num_magnitude_bins = 11
[
[(rotate, 0.5, 7), (solarize, 0.3, 2)],
[(shear_x, 0.2, 5), (posterize, 0.6, 8)],
[(rotate, 0.9, 6), (posterize, 0.1, 10)],
]
```

Here the samples processing is split in three paths. However images that fall into 1st and 3rd sub-policy can be run together in first stage. Similarily, in the second stage, the samples that fall into 2nd and 3rd policy can be run together. 

DALI will execute such policy by first: splitting batch into one that is omitted this time based on the probabilities from the first stage, i.e. `(0.5, 0.2 and 0.9)` and the one to process in the given stage, then the latter batch is split between rotate and shear_x. The rotate is called with per-sample angle that corresponds to the magnitudes from sub_policy (7 and 6). The batch is then merged and similar split is done for the second stage.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3193

### CODE REVIEW

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
!build

**Review by awolant — APPROVED** [review]
Well documented, well tested (+1 for using nose2 utils), very good descriptions of errors. Nice

File: dali/test/python/auto_aug/test_auto_augment.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        param_name = "max_translate_rel"
+    else:
+        param_name = "max_translate_abs"
+    if extent == 'both':
+        param = shape[0] * offset_fraction
+    if extent == 'height':
+        param = [shape[0] * offset_fraction, 0]
+    elif extent == 'width':
+        param = [0, shape[1] * offset_fraction]
+    params[param_name] = param
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/test/python/auto_aug/test_auto_augment.py:114]: Fixed

**Comment by stiepan** [comment]
!build

**Review by szalpal — APPROVED** [review]

File: dali/python/nvidia/dali/auto_aug/augmentations.py [hunk 2]
```diff
@@ -12,7 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import numpy as np
+try:
+    import numpy as np
+except ImportError:
+    raise RuntimeError(
+        "Could not import numpy. DALI's automatic augmentation examples depend on numpy. "
+        "Please install numpy to use the examples.")
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/python/nvidia/dali/auto_aug/augmentations.py:20]: I kind of followed a pattern here used elsewhere in DALI, but that makes a lot of sense. If we decide to change it, we can remove all such import wrapper at once.

**Pull request merged**